boards: arm: Add support for STM32L4R9I-DISCO

mathieuchopstm opened the pull request:
Used several existing STM32 boards (L476G, L496G, ...) as references for the board adding process.

The following features have been tested and confirmed working on hardware:
* I2C
* RTC
* SPI
* SDMMC
* UART
* OctoSPI Flash

gautierg-st commented on boards/arm/stm32l4r9i_disco/stm32l4r9i_disco.dts:
```diff
@@ -0,0 +1,272 @@
+#include "arduino_r3_connector.dtsi"
+
+/ {
+    model = "STMicroelectronics STM32L4R9I-DISCO board";
+    compatible = "st,stm32l4r9i-disco";
```
Fix indentation in this file. Use tab of size 8

gautierg-st reviewed (changes requested):
Some minor fixes. Please also fix the issues raised by CI.

erwango commented on boards/arm/stm32l4r9i_disco/stm32l4r9i_disco.dts:
```diff
@@ -0,0 +1,271 @@
+
+	leds {
+		compatible = "gpio-leds";
+		//N.B. LD1 (orange) is not wired to MCU
```
Please use C style comments.

erwango commented on boards/arm/stm32l4r9i_disco/stm32l4r9i_disco.dts:
```diff
@@ -0,0 +1,271 @@
+	 * WORKAROUND: stm32l4-pll-clock does not allow arbitrary PLLP dividers.
+	 * Disable PLLP completely since it only feeds SAI, which is not active either.
+	 */
+	//div-p = <5>;
```
Ditto

erwango commented on boards/arm/stm32l4r9i_disco/stm32l4r9i_disco.dts:
```diff
@@ -0,0 +1,271 @@
+
+	pwm3: pwm {
+		status = "okay";
+		//N.B.: Datasheet indicates that ARD_D11 (wired to PB15) is connected to TIM3_CH2.
+		//However, this is incorrest as PB15 cannot be muxed to TIM3 (see DS12023).
+		//Moved ARD_D11 to TIM15_CH2 instead.
```
Ditto

mathieuchopstm replied:
Done + fixed typo as well.

erwango commented on boards/arm/stm32l4r9i_disco/stm32l4r9i_disco.dts:
```diff
@@ -0,0 +1,271 @@
+
+&adc1 {
+	status = "okay";
+	pinctrl-0 = <&adc1_in5_pa0 &adc1_in12_pa7 &adc1_in15_pb0 &adc1_in4_pc3 &adc1_in13_pc4>;
```
Break a line:
```suggestion
	pinctrl-0 = <&adc1_in5_pa0 &adc1_in12_pa7 &adc1_in15_pb0 
	                  &adc1_in4_pc3 &adc1_in13_pc4>;
```

mathieuchopstm replied:
Done.

erwango commented on boards/arm/stm32l4r9i_disco/stm32l4r9i_disco.dts:
```diff
@@ -0,0 +1,271 @@
+		spi-bus-width = <OSPI_OPI_MODE>;
+		data-rate = <OSPI_STR_TRANSFER>;
+		four-byte-opcodes;
+		sfdp-bfp = [
```
Might not be required anymore after https://github.com/zephyrproject-rtos/zephyr/pull/62654

erwango commented on boards/arm/stm32l4r9i_disco/stm32l4r9i_disco_defconfig:
```diff
@@ -0,0 +1,20 @@
+CONFIG_CLOCK_CONTROL=y
+
+# Enable pin controller
+CONFIG_PINCTRL=y
```
Please add :
```
# Enable MPU
CONFIG_ARM_MPU=y
# Enable HW stack protection
CONFIG_HW_STACK_PROTECTION=y
```
Should also be added on the board you used as example if not already set (make it a dedicated commit)

mathieuchopstm replied:
Example boards have this config, I just didn't copy it as I didn't check if MCU had a MPU.
Done.

erwango commented on tests/drivers/adc/adc_api/boards/stm32l4r9i_disco.overlay:
```diff
@@ -0,0 +1,24 @@
+/*
```
Please split this in a dedicated commit.

mathieuchopstm replied:
Done.

mathieuchopstm committed:
tests: drivers/adc: Add STM32L4R9I-DISCO overlay

Adds the required overlay for test to build and
pass successfully on STM32L4R9I-DISCO board.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
boards: arm: Add support for STM32L4R9I-DISCO

Adds the device trees for the board and its Arduino connector,
default Kconfig and documentation.

The following features have been confirmed working on hardware:
* ADC
* I2C
* RTC
* SPI
* SDMMC
* UART
* OctoSPI Flash

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>
